monte_carlo.py

```diff
@@ -0,0 +1,195 @@
+import numpy as np
+from tqdm import tqdm
+from collections import defaultdict
+import warnings
+
+from cell import CELL
+
+
+class Policy:
+    def __init__(self, train_epsilon, test_epsilon):
+        self.policy = {}
+    
+        self.train_epsilon = train_epsilon
+        self.test_epsilon = test_epsilon
+        self.epsilon = train_epsilon
+
+    def __getitem__(self, state):
+        if state not in self.policy:
+            self.policy[state] = np.random.choice(MonteCarlo.get_legal_actions(state))
+        if np.random.random() < self.epsilon:
+            return np.random.choice(MonteCarlo.get_legal_actions(state))
+        return self.policy[state]
+    
+    def __setitem__(self, state, action):
+        self.policy[state] = action
+
+    def __len__(self):
+        return len(self.policy)
+    
+    def set_trained(self):
+        self.epsilon = self.test_epsilon
+
+    def load(self, filename):
+        with open(filename, 'r') as f:
+            for line in f:
+                state = tuple(map(int, line.split(',')[:4]))
+                action = line.split(',')[4].strip()
+                self.policy[state] = action
+
+    def save(self, filename):
+        with open(filename, 'w') as f:
+            for state, action in self.policy.items():
+                f.write(f"{state[0]},{state[1]},{state[2]},{state[3]},{action}\n")
+    
+class MonteCarlo:
+
+    actions = {'increase_vx': (1, 0), 'decrease_vx': (-1, 0),
+           'increase_vy': (0, 1), 'decrease_vy': (0, -1),
+           'increase_vx_vy': (1, 1), 'decrease_vx_vy': (-1, -1),
+           'increase_vx_decrease_vy': (1, -1), 'decrease_vx_increase_vy': (-1, 1),
+           'no_change': (0, 0)}
+    max_velocity = 2
+    
+    def __init__(self, grid, gamma=0.9, num_episodes=10000, train_epsilon=0.9, test_epsilon=0.05, policy_filename=None, load_policy=False):
+        self.grid = grid
+
+        self.starts = np.where(self.grid == CELL.START)
+        self.starts = list(zip(self.starts[0], self.starts[1]))
+        self.goal = np.where(self.grid == CELL.TARGET)
+        self.goal = (self.goal[0][0], self.goal[1][0])
+        self.obstacles = np.where(self.grid == CELL.WALL)
+        self.obstacles = list(zip(self.obstacles[0], self.obstacles[1]))
+
+        self.gamma = gamma
+        self.num_episodes = num_episodes
+        self.train_epsilon = train_epsilon
+        self.test_epsilon = test_epsilon
+        
+        self.Q = defaultdict(lambda: defaultdict(float))
+        self.returns = defaultdict(lambda: defaultdict(list))
+
+        self.policy = Policy(train_epsilon, test_epsilon)
+        self.policy_filename = policy_filename
+        self.is_trained = False
+
+        if load_policy:
+            self.policy.load(policy_filename)
+            self.is_trained = True
+            self.policy.set_trained()
+
+    @staticmethod
+    def get_legal_actions(state):
```

Comment:
add memoization for double calculation of similiar states?

Reply:
Again very difficult, as this is a static method in its current architecture.